pkg/engine/lifecycletest/alias_test.go

```diff
@@ -1648,3 +1648,102 @@ func TestSplitUpdateComponentAliases(t *testing.T) {
 	assert.Equal(t, resource.URN("urn:pulumi:test::test::pkgA:m:typB$pkgA:m:typC::resC"), snap.Resources[2].URN)
 	assert.Equal(t, resource.URN("urn:pulumi:test::test::pkgA:m:typB::resB"), snap.Resources[2].Parent)
 }
+
+func TestFailDeleteDuplicateAliases(t *testing.T) {
+	t.Parallel()
+
+	// This is a test for https://github.com/pulumi/pulumi/issues/14041 to check that we don't courrupt state files when
+	// old aliased resources are left in state because they can't delete.
+
+	mode := 0
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CreateF: func(urn resource.URN, news resource.PropertyMap, timeout float64,
+					preview bool,
+				) (resource.ID, resource.PropertyMap, resource.Status, error) {
+					// We should only create things in the first and last pass
+					ok := mode == 0 || mode == 2
+					assert.True(t, ok, "%s tried to create but should be aliased", urn)
+
+					return "created-id", news, resource.StatusOK, nil
+				},
+				DeleteF: func(urn resource.URN, id resource.ID, olds resource.PropertyMap,
+					timeout float64,
+				) (resource.Status, error) {
+					// We should only delete things in the last pass
+					ok := mode == 2
+					assert.True(t, ok, "%s tried to delete but should be aliased", urn)
+
+					return resource.StatusUnknown, errors.New("can't delete")
+				},
+			}, nil
+		}, deploytest.WithoutGrpc),
+	}
+
+	program := deploytest.NewLanguageRuntime(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		switch mode {
+		case 0:
+			fallthrough
+		case 2:
+			// Default case, and also the last case make "resA"
+			_, _, _, err := monitor.RegisterResource(
+				"pkgA:m:typA",
+				"resA",
+				true,
+				deploytest.ResourceOptions{})
+			assert.NoError(t, err)
+
+		case 1:
+			// Rename "resA" to "resAX" with an alias
+			_, _, _, err := monitor.RegisterResource(
```

Comment:
can't imagine that it would change anything, but should we verify the fix with a parent/child relationship where the parent is aliased?

Reply:
I think given the other fixes we've made to fixing up child URNs on snapshot write out, that would work as expected.

If you had "parent" and "child", then rename "parent" to "parentX" in a successful up all the URNs will be correct. If you then tried to go back to the program with "parent" and "child" (and no aliases) the engine would consider this two new resources and would try to delete "parentX" and reuse the "child" resource (because the URN would now be the same if the only difference between "parent" and "parentX" was the name). Even if "parentX" failed to delete in that pass it would just be left in the state file but not affecting anything else.